File: graph_search/145_binary_tree_postorder_traversal.cc
#include <iostream>
#include <stack>
#include <vector>

struct TreeNode {
  int value;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : value(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : value(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : value(x), left(left), right(right) {}
};

void Postorder(TreeNode* root, std::vector<int>& result) {
  if (root == nullptr) {
    return;
  }
  Postorder(root->left, result);
  Postorder(root->right, result);
  result.push_back(root->value);
}

std::vector<int> PostorderTraversal(TreeNode* root) {
  std::vector<int> result;
  Postorder(root, result);
  return result;
}

int main(int argc, char* argv[]) {
  TreeNode* node_1 = new TreeNode(1);
  TreeNode* node_2 = new TreeNode(2);
  TreeNode* node_3 = new TreeNode(3);
  node_1->right = node_2;
  node_2->left = node_3;
  std::vector<int> result = PostorderTraversal(node_1);
  for (int value : result) {
    std::cout << value << '\t';
  }
  std::cout << std::endl;
  delete node_1;
  delete node_2;
  delete node_3;
  return 0;
}

File: tree/ordered_traversal/145_binary_tree_postorder_traversal.py
from typing import List, Optional
from tree.tree_node import TreeNode, create_tree_node, print_tree_node


# 2026/05/19: 直接抄答案.
def postorder_traversal(self, root: Optional[TreeNode]) -> List[int]:
    res = []
    if root is None:
        return res

    stack = []
    prev = None
    stack.append(root)
    while root and len(stack) > 0:
        while root:
            stack.append(root)
            root = root.left
        root = stack.pop()
        if root.right is None or root.right == prev:
            res.append(root.val)
            prev = root
            root = None
        else:
            stack.append(root)
            root = root.right
    return res


root = [1, 2, 3, 4, 5, None, 8, None, None, 6, 7, 9]
